sdk/python/kfp/dsl/_pipeline_volume.py

```diff
@@ -57,8 +51,8 @@ def __init__(self,
             init_volume = {attr: getattr(volume, attr)
                            for attr in self.attribute_map.keys()}
         else:
-            init_volume = {"name": kwargs.pop("name")
-                           if "name" in kwargs else None}
+            init_volume = {"name": kwargs.pop("name") if "name" in kwargs
+                           else "pvolume-%s" % id(self)}
```

Comment:
It might be better to use a stable hash of serialized volume structure instead of id.

Reply:
The thing here is that we need to select a name in order to call `super().__init__()` which defines and fills the object's fields.
What does the structure contain before that? And if we do serialize the structure before the actual `init()`, wouldn't we get the same hash for every `PipelineVolume`? (since the input of the function will be the same _empty_ structure)

Reply:
I understand that the designs where the ID/name is part of the structure make hashing problematic.
What I did was setting an empty id/name during creation, then hashing the structure and then setting the id/name based on the hash. Do you think this is appropriate here? The advantage is that the compiled pipeline is not changing with every rebuild.

Reply:
I implemented your suggestion.
First, I used a placeholder for the `name`.
Then, I used the `to_dict()` method the Kubernetes model has (I think there is no need to convert it the same way the compiler does, i.e. proper yaml representation).
Finally, I hashed the `str()` representation of that dictionary.

What do you think?
Also, take a look at the test. Since the name is the same, I `assertEqual()` with the name it should have, hard-coded. Should I leave it like this, delete the test, or perform the same actions in the test function (create a volume with the placeholder name and hash it) and then compare them?

Reply:
The test will sometimes fail since python dict is not ordered, thus the str(self.to_dict()) is not deterministic.